pkg/ebpf/c/tracee.bpf.c

```diff
@@ -1965,40 +2005,154 @@ static __always_inline int init_tailcall_program_data(program_data_t *p, void *c
+{
+    // check equality_filter_matches() for more info
+
+    u64 equal_in_scopes = 0;
+    u64 set_in_scopes = 0;
+    eq_t *equality = bpf_map_lookup_elem(&binary_filter, proc_info->binary.path);
+    if (equality == NULL) {
+        // lookup by binary path and mount namespace
+        equality = bpf_map_lookup_elem(&binary_filter, &proc_info->binary);
+    }
+    if (equality != NULL) {
+        equal_in_scopes = equality->equal_in_scopes;
+        set_in_scopes = equality->set_in_scopes;
+    }
 
-    return filter_out;
+    return equal_in_scopes | (filter_out_scopes & ~set_in_scopes);
 }
 
-static __always_inline int equality_filter_matches(bool filter_out, void *filter_map, void *key)
+static __always_inline u64 equality_filter_matches(u64 filter_out_scopes,
+                                                   void *filter_map,
+                                                   void *key)
 {
-    u32 *equality = bpf_map_lookup_elem(filter_map, key);
-    if (equality != NULL)
-        return *equality;
+    // check compute_scopes() for initial info
+    //
+    // e.g.: cmdline: -t 2:comm=who -t 3:comm=ping -t 4:comm!=who
+    //
+    // filter_out_scopes = 0000 1000, since scope 4 has "not equal" for comm filter
+    // filter_map        = comm_filter
+    // key               = "who" | "ping"
+    //
+    // ---
+    //
+    // considering an event from "who" command
+    //
+    // equal_in_scopes   = 0000 0010, since scope 2 has "equal" for comm filter
+    // set_in_scopes     = 0000 1010, since scope 2 and 4 are set for comm filter
+    //
+    // return            = equal_in_scopes | (filter_out_scopes & ~set_in_scopes)
+    //                     0000 0010 |
+    //                     (0000 1000 & 1111 0101) -> 0000 0000
+    //
+    //                     0000 0010 |
+    //                     0000 0000
+    //                     ---------
+    //                     0000 0010 = (scope 2 matched)
+    //
+    // considering an event from "ping" command
+    //
+    // equal_in_scopes   = 0000 0100, since scope 3 has "equal" for comm filter
+    // set_in_scopes     = 0000 0100, since scope 3 is set for comm filter
+    //
+    // return            = equal_in_scopes | (filter_out_scopes & ~set_in_scopes)
+    //                     0000 0100 |
+    //                     (0000 1000 & 0000 0100) -> 0000 0000
+    //
+    //                     0000 0100 |
+    //                     0000 0000
+    //                     ---------
+    //                     0000 0100 = (scope 3 matched)
 
-    return filter_out;
+    u64 equal_in_scopes = 0;
+    u64 set_in_scopes = 0;
+    eq_t *equality = bpf_map_lookup_elem(filter_map, key);
+    if (equality != NULL) {
+        equal_in_scopes = equality->equal_in_scopes;
+        set_in_scopes = equality->set_in_scopes;
+    }
+
+    return equal_in_scopes | (filter_out_scopes & ~set_in_scopes);
```

Comment:
> Need to check when we use the != operator (since we usually test the = operator)

**-t 2:comm=ls**

config_entry:

```
"comm_filter_enabled_scopes": 2
"comm_filter_out_scopes": 0
```

comm_filter:

```
[{
        "key": {
            "str": "ls"
        },
        "value": {
            "equal_in_scopes": 2,
            "enabled_in_scopes": 2
        }
    }
]
```

```c
return equal_in_scopes | (filter_out_scopes & ~enabled_in_scopes);
 2 | (0 & ~2) = 2 (0010)
```

**-t 2:comm!=ls**

config_entry:

```
"comm_filter_enabled_scopes": 2
"comm_filter_out_scopes": 2
```

comm_filter:

```
[{
        "key": {
            "str": "ls"
        },
        "value": {
            "equal_in_scopes": 0,
            "enabled_in_scopes": 2
        }
    }
]
```

```c
return equal_in_scopes | (filter_out_scopes & ~enabled_in_scopes);
 0 | (2 & ~2) = ..FD (1101)
```

Reply:
So what's the conclusion? To me it seems that we indeed need to change to `filter_out_scopes & enabled_in_scopes);` so we will get `2 & 2` in the last example

Reply:
After tests with `return equal_in_scopes | (filter_out_scopes & enabled_in_scopes);` we got no regressions:

with equal and non equal to the same event filter 

`1:event!=accept4 -t 1:pid=1`

0000000000000001 scope 1 for all events (but accept4) from pid 1

`-t 42:event=accept4 -t 42:pid=1 -t 10:event=accept4 -t 10:pid=1`
 
0000020000000200 scope 42 and 10 for accept4 events from pid 1

![image](https://user-images.githubusercontent.com/3372117/213719315-534f56c6-c7c6-4107-8052-59a4aa0bb182.png)

Then, changing it to `return equal_in_scopes | (filter_out_scopes & enabled_in_scopes);`.

Reply:
> we got no regressions

This statement is wrong. In other test cases, we caught errors, which conclude that the current proposed logic is correct. No changes required.